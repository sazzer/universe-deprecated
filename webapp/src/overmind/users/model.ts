/** The state of the user record */
export type UserState = "LOADING" | "SAVING" | "UNKNOWN" | "LOADED";

/** The shape of a single user */
export interface User {
  /** The unqiue ID of the user */
  id: string;
  /** The unique username of the user */
  username: string;
  /** The display name of the user */
  displayName: string;
  /** The email address of the user, if we know what it is */
  email: string | null;
}

/** Representation of a user as known by the state */
export interface StoredUser {
  /** The user record, if we've currently got one */
  user: User | undefined;

  /** The current state of the user */
  state: UserState;
}

/** Error when loading a user that doesn't exist */
export class UnknownUserError extends Error {}
