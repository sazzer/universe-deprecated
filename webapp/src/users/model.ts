/**
 * Representation of the user that is currently using the application
 */
export interface User {
  /** The User ID of the user */
  id: string;
  /** The username of the user */
  username: string;
  /** The display name of the user */
  displayName: string;
  /** The email address of the user */
  email: string;
}

/**
 * Representation of an Access Token for a user
 */
export interface AccessToken {
  /** The actual token */
  token: string;
  /** When the token expires */
  expiry: string;
}

/**
 * Representation of a User that has just authenticated, and thus also has an access token
 */
export interface AuthenticatedUser extends User {
  /** The access token for the user */
  accessToken: AccessToken;
}
