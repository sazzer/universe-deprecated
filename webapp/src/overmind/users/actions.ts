import { Action } from "overmind";
import { User } from "./model";

/**
 * Overmind action for storing the details of a user
 */
export const storeUser: Action<User> = ({ state }, details: User) => {
  state.users.users[details.id] = {
    id: details.id,
    username: details.username,
    displayName: details.displayName,
    email: details.email
  };
  delete state.users.userStates[details.id];
};

export const fetchUser: Action<string, Promise<void>> = async (
  { state, actions, effects },
  userId: string
) => {
  state.users.userStates[userId] = "LOADING";
  const user = await effects.users.api.loadUser(userId);
  actions.users.storeUser(user);
};
