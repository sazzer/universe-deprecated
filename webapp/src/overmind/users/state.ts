export interface User {
  userId: string;
  username: string;
  displayName: string;
  email: string | null;
}

type State = {
  users: User[];
};

export const state: State = {
  users: []
};
