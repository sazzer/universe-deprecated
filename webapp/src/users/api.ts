import { ProblemResponse, request } from "../api";

import { AuthenticatedUser } from "./model";
import debug from "debug";

/** The logger to use */
const LOG = debug("universe:users:api");

/**
 * Check if a given username is registered with the service or not
 * @param username The username to check
 * @return True if the username is already registered. False if it's not.
 */
export async function checkUsername(username: string): Promise<boolean> {
  LOG("Checking the existance of username: %s", username);
  try {
    await request({
      url: "/usernames/{username}",
      urlParams: {
        username
      },
      method: "GET"
    });

    LOG("Existance of username %s: true", username);
    return true;
  } catch (e) {
    if (
      e instanceof ProblemResponse &&
      e.problem.type === "tag:universe,2020:users/problems/unknown-user"
    ) {
      LOG("Existance of username %s: false", username);
      return false;
    } else {
      LOG("Error checking existance of username %s: %o", username, e);
      throw e;
    }
  }
}

/**
 * Attempt to authenticate with the given credentials
 * @param username the username to authenticate with
 * @param password the password to authenticate with
 */
export async function authenticate(
  username: string,
  password: string
): Promise<AuthenticatedUser> {
  LOG("Authenticating as username %s with password %s", username, password);
  try {
    const user = await request<AuthenticatedUser>({
      url: "/login",
      method: "POST",
      data: {
        username,
        password
      }
    });

    LOG("Authenticated successfully: %o", user);
    return user.data;
  } catch (e) {
    LOG("Failed to authenticate as username %s: %o", username, e);
    throw e;
  }
}
