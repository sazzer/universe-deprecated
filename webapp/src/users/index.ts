export * from "./model";
export * from "./userContext";
export * from "./api";
