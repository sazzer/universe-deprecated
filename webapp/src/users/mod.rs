mod model;
mod service;

pub use model::*;
