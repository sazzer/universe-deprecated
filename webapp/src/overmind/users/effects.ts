import { request, ProblemResponse } from "../../api";
import { User, UnknownUserError } from "./model";

/**
 * Retrieve a user from the API
 * @param id The ID of the user to retrieve
 * @return The details of the user
 */
async function loadUser(id: string): Promise<User> {
  try {
    const user = await request<User>({
      url: "/users/{id}",
      urlParams: {
        id
      },
      method: "GET"
    });

    return user.data;
  } catch (e) {
    throw new UnknownUserError();
  }
}

/** The API for working with users*/
export const api = {
  loadUser
};
