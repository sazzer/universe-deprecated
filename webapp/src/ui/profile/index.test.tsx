import { getUserById, useUser } from "../../users";
import { render, wait } from "@testing-library/react";

import { ProfilePage } from "./index";
import React from "react";
import { StaticRouter as Router } from "react-router-dom";

jest.mock("../../users");

const useUserMock = useUser as jest.Mock;
const getUserByIdMock = getUserById as jest.Mock;

afterEach(() => {
  jest.resetAllMocks();
});

test("Rendering when not logged in", () => {
  useUserMock.mockReturnValue({
    hasUser: false,
    user: null,
    clearUser: () => {},
    storeUser: () => {}
  });

  const { container } = render(
    <Router location="/profile">
      <ProfilePage />
    </Router>
  );
  expect(container).toMatchSnapshot();
});

test("Rendering when logged in", async () => {
  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser: () => {}
  });

  getUserByIdMock.mockResolvedValueOnce({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });

  const { container } = render(
    <Router location="/profile">
      <ProfilePage />
    </Router>
  );

  // Waiting on getUserByIdMock
  await wait(() => {});

  expect(container).toMatchSnapshot();
});
