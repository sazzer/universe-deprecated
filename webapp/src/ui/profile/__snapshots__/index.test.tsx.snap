// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering when logged in 1`] = `
<div>
  <div
    data-test="profilePage"
  >
    <nav
      aria-label="breadcrumb"
    >
      <ol
        class="breadcrumb"
      >
        <li
          class="breadcrumb-item"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          aria-current="page"
          class="breadcrumb-item active"
        >
          User Profile
        </li>
      </ol>
    </nav>
    <div
      class="row"
    >
      <div
        class="col-12 col-md-9 order-sm-1"
      >
        <div
          class="tab-content"
        >
          <div
            aria-labelledby="profile-tabs-profile-tab"
            id="profile-tabs-profile"
            role="tabpanel"
          >
            <form
              data-test="userProfileForm"
            >
              <div
                class="form-group"
                data-test="username"
              >
                <label
                  for="username"
                >
                  Username
                </label>
                <input
                  class="form-control"
                  id="username"
                  name="username"
                  readonly=""
                  type="text"
                />
              </div>
              <div
                class="form-group"
                data-test="email"
              >
                <label
                  for="email"
                >
                  Email Address
                </label>
                <input
                  class="form-control"
                  id="email"
                  name="email"
                  type="text"
                />
              </div>
              <div
                class="form-group"
                data-test="displayName"
              >
                <label
                  for="displayName"
                >
                  Display Name
                </label>
                <input
                  class="form-control"
                  id="displayName"
                  name="displayName"
                  type="text"
                />
              </div>
              <div
                class="form-group"
              >
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  Save Changes
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
      <div
        class="col-12 col-md-3"
      >
        <div
          aria-orientation="vertical"
          class="nav flex-column nav-pills"
          id="profileSupportedContent"
          role="tablist"
        >
          <a
            aria-controls="profile-tabs-profile"
            aria-current="page"
            aria-selected="true"
            class="nav-link active"
            href="/profile"
            id="profile-tabs-profile-tab"
            role="tab"
          >
            User Profile
          </a>
          <a
            aria-controls="profile-tabs-password"
            aria-selected="false"
            class="nav-link"
            href="/profile/password"
            id="profile-tabs-password-tab"
            role="tab"
          >
            Change Password
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Rendering when not logged in 1`] = `
<div>
  You must be logged in to access this page
</div>
`;
